Tic-Tac-Toe-Python/TicTacToe.py

```diff
@@ -0,0 +1,140 @@
+                                       range(self._NUMBER_OF_ROWS)]  # initializing grid with empty string
+
+        self._player_1 = player1
+        self._player_2 = player2
+        self._is_current_player_1: bool = True
+
+    def _show_scores(self) -> None:
+        print(
+            f"Scores: \n {self._player_1.player_id}: {self._player_1.get_score()}\n {self._player_2.player_id}: {self._player_2.get_score()}")
+
+    def _show_grid(self) -> None:
+        for i in range(self._NUMBER_OF_ROWS):
+            print(' ', end='')
+            for j in range(self._NUMBER_OF_COLUMNS):
+                print('---', end=' ')
+            print('\n| ', end='')
+            print(' | '.join(self._grid[i]), end=' |\n')
+
+        print(' ', end='')
+
+        for j in range(self._NUMBER_OF_COLUMNS):
+            print('---', end=' ')
+        print('\n', end='')
+
+    def _new_game(self) -> None:
+        """
+        Sets grid to empty and current player to 1
+        :return:
+        """
+        self._grid: List[list] = [[' '] * self._NUMBER_OF_COLUMNS for _ in
+                                  range(self._NUMBER_OF_ROWS)]
+        self._is_current_player_1: bool = True
+
+    def _get_current_player(self) -> TicTacToeUserPlayer:
+        return self._player_1 if self._is_current_player_1 else self._player_2
+
+    def _check_win(self) -> bool:
+        """
+        Returns if the current player has won
+        """
+        main_diagonal = list()
+        alternate_diagonal = list()
+
+        for i in range(self._NUMBER_OF_ROWS):
+            # checking row
+            if self._grid[i].count(self._player_2.player_id) == self._NUMBER_OF_COLUMNS:
+                return True
+
+            if self._grid[i].count(self._player_1.player_id) == self._NUMBER_OF_COLUMNS:
+                return True
+
+            # checking ith column
+            ith_column = [self._grid[j][i] for j in range(self._NUMBER_OF_ROWS)]
+
+            if ith_column.count(self._player_1.player_id) == self._NUMBER_OF_COLUMNS:
+                return True
+
+            if ith_column.count(self._player_2.player_id) == self._NUMBER_OF_COLUMNS:
+                return True
+
+            main_diagonal.append(self._grid[i][i])
+            alternate_diagonal.append(self._grid[i][self._NUMBER_OF_ROWS - i - 1])
+
+        if main_diagonal.count(self._player_1.player_id) == self._NUMBER_OF_COLUMNS:
+            return True
+
+        if main_diagonal.count(self._player_2.player_id) == self._NUMBER_OF_COLUMNS:
+            return True
+
+        if alternate_diagonal.count(self._player_1.player_id) == self._NUMBER_OF_COLUMNS:
+            return True
+
+        if alternate_diagonal.count(self._player_2.player_id) == self._NUMBER_OF_COLUMNS:
+            return True
+
+        return False
+
+    def _check_game_over(self, move_count: int, total_moves: int) -> bool:
+        return move_count < total_moves
+
+    def play(self) -> None:
```

Comment:
this lets you play multiple games. A good thing here would be to have a method for a single game which is being used by a multiple game function . Also a multiple game may want to use a different TicTacToeGame Altogether. for example first game may be 3x3 but next might be 5x5 which will have you chnage the game object altogether.

```
@classmethod
def new(rows, cols):
     return TicTacToeGame(rows, cols)

def play(self):
    pass
```

and then in main

you can do 
```
while new_game:
      game = TicTacToeGame.new()
      game.play()
```